### PR TITLE
Corrects issue with phantom drawer displaying on a language change

### PR DESCRIPTION
**Pull request by brian-learningpool**
This appears to be because the drawer listens to the "adapt:initialize"
before checking that it has no items and should be hidden.

Since "adapt:initialize" should not be triggered more than once, I've added
a 'forceCheckVisibility' flag which forcees this check when the language
changes.

### CODE REVIEW

**Comment by brian-learningpool** [comment]
Resolves #1294.

File: src/core/js/views/drawerView.js [hunk 1]
```diff
@@ -8,8 +8,10 @@ define(function(require) {
         className: 'drawer display-none',
         disableAnimation: false,
         escapeKeyAttached: false,
+        forceCheckVisibility: false,
```
Comment (moloko) [inline src/core/js/views/drawerView.js]: could just use a comment explaining why this is necessary

Comment (moloko) [inline src/core/js/views/drawerView.js]: also `forceCheckAvailability` would work better as a name given that its purpose is to trigger a call to `checkIfDrawerIsAvailable`

File: src/core/js/views/drawerView.js [hunk 2]
```diff
@@ -78,6 +80,10 @@ define(function(require) {
         // Set tabindex for select elements
         postRender: function() {
             this.$('a, button, input, select, textarea').attr('tabindex', -1);
+
+            if (this.forceCheckVisibility) {
+                this.checkIfDrawerIsAvailable();
```
Comment (moloko) [inline src/core/js/views/drawerView.js]: out of interest, what happens if you always run `this.checkIfDrawerIsAvailable()` here? i.e. ditch doing it on `adapt:initialize`?

File: src/core/js/views/drawerView.js [hunk 3]
```diff
@@ -27,7 +27,7 @@ define(function(require) {
         setupEventListeners: function() {
             this.listenTo(Adapt, 'navigation:toggleDrawer', this.toggleDrawer);
             this.listenTo(Adapt, 'drawer:triggerCustomView', this.openCustomView);
-            this.listenToOnce(Adapt, 'adapt:initialize', this.checkIfDrawerIsAvailable);
+            this.listenTo(Adapt, 'app:dataReady', this.checkIfDrawerIsAvailable);
```
Comment (moloko) [inline src/core/js/views/drawerView.js]: now that `checkIfDrawerIsAvailable` has been added to `postRender` it actually works fine for me without this line

**Comment by moloko** [comment]
Annoyingly there is a slight flash of the resources icon when you change languages in Firefox. seems fine in IE11, Chrome and iOS10 though.
I suppose the only way to fix that would be to reverse the logic i.e. the icon should be hidden unless there are resources rather than shown unless there aren't.
It doesn't seem that serious if it's just FF though.

**Comment by brian-learningpool** [comment]
@moloko, I've updated the code based on your feedback.  When testing with Vanilla though, as the navigation template is overridden it requires a `display-none` added to the Drawer button to get rid of the flash here:
https://github.com/adaptlearning/adapt-contrib-vanilla/blob/master/templates/navigation.hbs#L5

I'm happy to make this change also, but if the Vanilla theme is released prior to the framework there will be an issue with the drawer not displaying.

File: src/core/js/views/drawerView.js [hunk 4]
```diff
@@ -9,7 +9,7 @@ define(function(require) {
         disableAnimation: false,
         escapeKeyAttached: false,
 
-        initialize: function() {
+        initialize: function(options) {
```
Comment (moloko) [inline src/core/js/views/drawerView.js]: the `options` parameter is now redundant I think?

File: src/core/js/app.js [hunk 5]
```diff
@@ -111,6 +112,9 @@ require([
     }
 
     function mapAdaptIdsToObjects () {
+        // Clear any existing mappings, as they are no longer valid.
+        Adapt.mappedIds = {};
```
Comment (oliverfoster) [inline src/core/js/app.js]: mapAdaptIdsToObjects  and mappedIds are unrelated. this should go into Adapt.setupMapping

mapAdaptIdsToObjects : is for indexing items (components, blocks, articles) in their collections
mappedIds: is for indexing the collections

Comment (brian-learningpool) [inline src/core/js/app.js]: Good catch. I've just moved it.

**Comment by moloko** [comment]
all works for me, get rid of that `options` parameter from `drawerView.initialize` and I'll plus-one

**Comment by oliverfoster** [comment]
i'll let you delete the branch

**Pull request merged**